tests/fs/api.private.test.ts

```diff
@@ -0,0 +1,284 @@
+import * as fc from 'fast-check'
+
+import { IPFS } from 'ipfs-core'
+import { createInMemoryIPFS } from '../helpers/in-memory-ipfs'
+import FileSystem from '../../src/fs/filesystem'
+
+import * as ipfsConfig from '../../src/ipfs'
+import * as path from '../../src/path'
+import * as crypto from '../../src/crypto'
+
+
+let ipfs: IPFS | null = null
+let rootKey: string
+
+beforeAll(async () => {
+  ipfs = await createInMemoryIPFS()
+  ipfsConfig.set(ipfs)
+})
+
+afterAll(async () => {
+  if (ipfs === null) return
+  await ipfs.stop()
+})
+
+fc.configureGlobal({ numRuns: 10 })
+
+describe('the filesystem api', () => {
+  it('writes files', async () => {
+    const fs = await emptyFilesystem()
+
+    await fc.assert(
+      fc.asyncProperty(
+        fc.tuple(pathSegment(), fileContent()), async data => {
```

Comment:
`data` seems like a really generic function name. Also, let's say (hypothetically) you'd at some point change an `fc.tuple` call from 3 arguments to 2 arguments. It might be pretty easy to miss an undefined `data[2]` access.

What do you think about using pattern matching here?
```ts
 fs.tuple(pathSegment(), fileContent()), async ([pathSegment, [val, type]]) => { // probably needs a type signature
    // ...
}
```

I don't really know what the types are, but it should be something like `[pathSegment, [val, type]]: [string[], [<sth>, <sth>]]`.

Reply:
I've been playing with this idea and `fc.record`. Both suggestions are great, thanks for calling them out. :pray: Will clean up these tests nicely.

I'm leaning towards tuples with pattern matching over records. The records are nice because they are more constrained, but I'm not quite sure what to with the tuples in `fileContent` and `pathSegmentPair`. It seems like we would need to destructure or index in the test body for these. 

Tuples with pattern matching seems a bit more readable to me. I'll change over to using that, but I'm open to discussing records more.

Reply:
Can we perhaps make `fileContent` (and possible `pathSegmentPair`) defined using `fc.record`, too? (I hope I'm understanding you correctly here)

Reply:
Yes, you are. And it's a good suggestion. 

`pathSegmentPair` is a bit trickier than `fileContent` because it's a set, not a tuple:

```javascript
const pathSegmentPair = () => {
  return fc.set(
    fc.hexaString({ minLength: 4, maxLength: 20 }),
    { minLength: 2, maxLength: 2 }
  )
}
```

but we might be able to get the same effect using `filter` on records. Also, the minimum length of 4 keeps the members of the pair unique in most cases. :thinking:

Reply:
Ah right. Yeah let's keep that an `fc.set`.

> It seems like we would need to destructure or index in the test body for these.

Keep in mind that you can also destructure record fields in javascript (unlike in e.g. Elm):
```js
const { pathPair: [first, second] } = { pathPair: ["abc", "def"] }
```

---

I'm fine either way. I see it might be more involved. I'll leave it to your judgement :)

Reply:
Oh, one more option: There's a `.map` for `Arbitrary`s in fast-check. So this would work:

```js
const pathSegmentPair = () => {
  return fc.set(
    fc.hexaString({ minLength: 4, maxLength: 20 }),
    { minLength: 2, maxLength: 2 }
  ).map(([first, second]) => ({ first, second }))
}
```

In case that'd make it any more readable :sweat_smile:

Reply:
It's great! Fun to explore the possibilities here and learn more `fast-check` along the way. :smiley:

Reply:
I went ahead with records instead of tuples. With a bit of pattern matching, it came out fairly readable.

Thanks again for all the great suggestions on this. :pray: